oauth/src/lib.rs

```diff
@@ -0,0 +1,227 @@
+}
+
+#[derive(Debug)]
+pub struct OAuthToken {
+    pub access_token: String,
+    pub refresh_token: String,
+    pub expires_at: Instant,
+    pub token_type: String,
+    pub scopes: Vec<String>,
+}
+
+/// Return code query-string parameter from the redirect URI.
+fn get_code(redirect_url: &str) -> Result<AuthorizationCode, OAuthError> {
+    let url = Url::parse(redirect_url).map_err(|e| OAuthError::AuthCodeBadUri {
+        uri: redirect_url.to_string(),
+        e,
+    })?;
+    let code = url
+        .query_pairs()
+        .find(|(key, _)| key == "code")
+        .map(|(_, code)| AuthorizationCode::new(code.into_owned()))
+        .ok_or(OAuthError::AuthCodeNotFound {
+            uri: redirect_url.to_string(),
+        })?;
+
+    Ok(code)
+}
+
+/// Prompt for redirect URI on stdin and return auth code.
+fn get_authcode_stdin() -> Result<AuthorizationCode, OAuthError> {
+    println!("Provide redirect URL");
+    let mut buffer = String::new();
+    let stdin = io::stdin();
+    stdin
+        .read_line(&mut buffer)
+        .map_err(|_| OAuthError::AuthCodeStdinRead)?;
+
+    get_code(buffer.trim())
+}
+
+/// Spawn HTTP server on provided socket to accept OAuth callback and return auth code.
+fn get_authcode_listener(socket_address: SocketAddr) -> Result<AuthorizationCode, OAuthError> {
+    let listener =
+        TcpListener::bind(socket_address).map_err(|e| OAuthError::AuthCodeListenerBind {
+            addr: socket_address,
+            e,
+        })?;
+    info!("OAuth server listening on {:?}", socket_address);
+
+    // The server will terminate itself after collecting the first code.
+    let mut stream = listener
+        .incoming()
+        .flatten()
+        .next()
+        .ok_or(OAuthError::AuthCodeListenerTerminated)?;
+    let mut reader = BufReader::new(&stream);
+    let mut request_line = String::new();
+    reader
+        .read_line(&mut request_line)
+        .map_err(|_| OAuthError::AuthCodeListenerRead)?;
+
+    let redirect_url = request_line
+        .split_whitespace()
+        .nth(1)
+        .ok_or(OAuthError::AuthCodeListenerParse)?;
+    let code = get_code(&("http://localhost".to_string() + redirect_url));
+
+    let message = "Go back to your terminal :)";
+    let response = format!(
+        "HTTP/1.1 200 OK\r\ncontent-length: {}\r\n\r\n{}",
+        message.len(),
+        message
+    );
+    stream
+        .write_all(response.as_bytes())
+        .map_err(|_| OAuthError::AuthCodeListenerWrite)?;
+
+    code
+}
+
+// TODO: Pass in redirect_address instead since the redirect host depends on client ID?
+/// Obtain a Spotify access token using the authorization code with PKCE OAuth flow.
+pub fn get_access_token(
+    client_id: &str,
+    scopes: Vec<&str>,
+    redirect_port: Option<u16>,
+) -> Result<OAuthToken, OAuthError> {
+    // Must use host 127.0.0.1 with Spotify Desktop client ID.
+    let redirect_address = SocketAddr::new(
+        IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
```

Comment:
Mindfart: how does that work for IPv6-only hosts? If Spotify supports that at all?

Reply:
I could certainly try setting the redirect host ip to ::1 and test if they accept that, my random guess is they won't, buy it's quick enough to try. As it is, someone running ipv6-only would have to use the stdin version of the code recovery as  a workaround (`--oauth-port 0`). I don't have an ipv6 only system to test with so I'm not that keen on trying to implement it.

Reply:
I agree, just a mindfart triggered by #1307. Don't worry about it.

Reply:
Using ipv6 localhost in the `redirect_uri` param for the Oauth flow did actually work so I think it would be possible to support this properly... but I'd still rather leave it for someone who could test it properly.

Reply:
@roderickvd I fixed the last todo so now takes a `redirect_uri` argument instead of just the port. If you set that to http://[::1]/foo etc, you'll get an IPv6 socket.